mypy/build.py

```diff
@@ -1011,8 +1012,8 @@ def read_protocol_cache(manager: BuildManager,
     current_meta_snapshot = {}  # type: Dict[str, str]
     for id in graph:
         meta = graph[id].meta
-        assert meta is not None, 'Protocol cache should be read after all other'
-        current_meta_snapshot[id] = meta.hash
+        if meta:
```

Comment:
I am a bit worried about this change. My understanding is a following: if I first run mypy with only `--incremental` protocol cache is not written, if I then change some files and run again with `--incremental` and `--cache-fine-grained` it should still revert to a full run. In general, the caching logic is quite complex, and I think we need to first add tests that full run actually happens if:
* the above scenario happens
* protocol caches are inconsistent (we have wrong source hashes, because the protocol cache was copied from somewhere).

Reply:
I've manually tested those but will try to get a real test set up. Unfortunately our testing setup is really not good at this sort of thing.

The first case works fine in a pretty natural way without actually needing to trigger the retry logic really: every cache file fails to load because it doesn't have a fine-grained deps cache.